src/handlers.cpp

```diff
@@ -2529,59 +2522,171 @@ void PushProfileTask::on_get_primary_impus_success(CassandraStore::Operation* op
+  }
+  else
+  {
+    no_ims_set_first_default();
+    get_registration_set();
+  }
+}
 
-    if (!impu_match_found)
+bool PushProfileTask::default_ids_match()
+{
+  // Attempt to find a match between any of the default ids of the sets the
+  // private id corresponds to, and the default id in the PPR. If no match is
+  // found the PPR is changing the default id (which is not permitted), so
+  // reject it. Otherwise, continue.
+  XmlUtils::get_default_id(_ims_subscription, _new_default_id);
+  bool impu_match_found = false;
+  for (std::string current_default_id : _default_impus)
+  {
+    if (current_default_id == _new_default_id)
     {
-      TRC_INFO("The default id of the PPR doesn't match a default id already "
-               "known be belong to the IMPI %s - reject the PPR", _impi.c_str());
-      SAS::Event event(this->trail(), SASEvent::PPR_CHANGE_DEFAULT_IMPU, 0);
-      event.add_var_param(_impi);
-      event.add_var_param(new_default_id);
-      SAS::report_event(event);
-      send_ppa(DIAMETER_REQ_FAILURE);
-      return;
+      impu_match_found = true;
+      break;
     }
   }
+  return impu_match_found;
+}
 
-  // If we've reached this point, we can safely update the reg data in the
-  // knowledge it will not change the default public identity.
-  update_reg_data();
+void PushProfileTask::no_default_id_match_send_failure()
+{
+  TRC_INFO("The default id of the PPR doesn't match a default id already "
+           "known to belong to the IMPI %s - reject the PPR", _impi.c_str());
+  SAS::Event event(this->trail(), SASEvent::PPR_CHANGE_DEFAULT_IMPU, 0);
+  event.add_var_param(_impi);
+  event.add_var_param(_new_default_id);
+  SAS::report_event(event);
+  send_ppa(DIAMETER_REQ_FAILURE);
+  delete this;
+}
+
+void PushProfileTask::ims_sub_get_ids()
+{
+  _impus = XmlUtils::get_public_ids(_ims_subscription);
+
+  // We should check the IRS contains a SIP URI and throw an error log if
+  // it doesn't. We continue as normal even if it doesn't.
+  bool found_sip_uri = false;
+
+  for (std::vector<std::string>::iterator it = _impus.begin();
+      (it != _impus.end()) && (!found_sip_uri);
+       ++it)
+  {
+    if ((*it).compare(0, SIP_URI_PRE.length(), SIP_URI_PRE) == 0)
+    {
+      found_sip_uri = true;
+    }
+  }
+
+  if (!found_sip_uri)
+  {
+    TRC_ERROR("No SIP URI in Implicit Registration Set");
+    SAS::Event event(this->trail(), SASEvent::NO_SIP_URI_IN_IRS, 0);
+    event.add_compressed_param(_ims_subscription, &SASEvent::PROFILE_SERVICE_PROFILE);
+    SAS::report_event(event);
+  }
+
+  XmlUtils::get_default_id(_ims_subscription, _default_public_id);
+  _first_default_id = _default_public_id;
+  _default_impus.erase(std::remove(_default_impus.begin(),
+                                   _default_impus.end(),
+                                   _default_public_id),
+                       _default_impus.end());
+  _default_impus.push_back(_first_default_id);
+}
+
+void PushProfileTask::no_ims_set_first_default()
```

Comment:
I'm not sure this warrants being in its own function - I think the code flow would be easier to understand if it was just these two lines called directly (with an appropriate comment).